---
title: "Methrix tutorial"
author: "CompEpigen"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
    self_contained: yes
    highlight: pygments
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Introduction
This vignette describes basic usage of the package intended to process several large [bedgraph](https://genome.ucsc.edu/goldenPath/help/bedgraph.html) files in R. `Methrix` provides set of function which allows easy importing of various flavours of bedgraphs generated by methylation callers, and many downstream analysis to be performed on large matrices.

## Reading bedgraph files
`read_bedgraphs` function is a versatile bedgraph reader intended to import bedgraph files generated virtually by any sort of methylation calling program. It requires user to provide indices for chromosome names, start position and other required fields. There are also presets available to import `bedgraphs` from most common programs such as `Bismark`, `MethylDackel`, and `MethylcTools`.

```{r, message=FALSE, warning=FALSE}
#Load library
library(methrix)
library(BSgenome.Hsapiens.1000genomes.hs37d5) #Genome of your preference to work with
```


```{r}
#Get the list of files to process
bed_files = list.files(path = "~/C010-Datasets/Internal/2019-03-25_JMMLC_PBAT/JMML_Methylation/01_raw_data/bedgraphs/CG/", 
                       pattern = "*.bed.gz", full.names = TRUE)
#Normalize the path
bed_files = normalizePath(path = bed_files)
#For vignette purpose use three files
bed_files = bed_files[1:3]

print(bed_files)
```

We can import bedgraph files with the function `read_bedgraphs` which reads in the bedgraphs, adds CpGs missing from the reference set, and creates a methylation/coverage matrices. If your system has enough memory increasing `vect_batch_size` and `n_threads` substantially speeds up the process. Once the process is complete - it returns an object of class `methrix` which in turn inherits [SummarizedExperiment](https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html) class. `methrix` object contains 'methylation' and 'coverage' matrices (either in-memory or as on-disk HDF5 arrays) along with pheno-data and other basic info. This object can be passed to all downstream functions for various analysis.

Bedgraphs used for vignette are generated with `MethylcTools` from bam files aligned to [GRCh37-1000G](http://lh3.github.io/2017/11/13/which-human-reference-genome-to-use) reference genome. 

```{r, warning=FALSE}
#First extract genome wide CpGs from the desired genome
hs37d5_cpgs = methrix::extract_CPGs(ref_genome = "BSgenome.Hsapiens.1000genomes.hs37d5", bored = FALSE)
```

```{r}
#Read the files
meth = methrix::read_bedgraphs(files = bed_files, 
                               pipeline = "MethylcTools",
                               stranded = TRUE,
                               collapse_starnds = TRUE, 
                               ref_build = "Hs37d5", 
                               vect_batch_size = 3, 
                               ref_cpgs = hs37d5_cpgs)
```

```{r}
#Typing meth shows basic summary.
meth
```

## Basic operations
```{r}
#Get chromosome wise number of CpGs
getChrSummary(x = meth)
```

Its possible to extract methylation/coverage matrices using `get_matrix` function.
```{r}
#Coverage matrix
coverage_mat = methrix::get_matrix(m = meth, type = "C")
head(coverage_mat)
```

```{r}
#Methylation matrix
meth_mat = methrix::get_matrix(m = meth, type = "M")
head(meth_mat)
```

```{r}
#If you prefer you can attach loci info to the matrix
meth_mat_with_loci = methrix::get_matrix(m = meth, type = "M", add_lcoi = TRUE)
meth_mat_with_loci
```

## Filtering
### Remove uncovered loci
Usual task in analysis involves removing uncovered CpGs. i.e, those loci which are not covered across all sample (in other words covered only in subset of samples resulting `NA` for rest of the samples ).
```{r}
meth_clean = methrix::remove_uncovered(m = meth)
meth_clean
```

### Coverage filter
Furthermore if you prefer you can filter sites based on coverage conditions.
```{r}
#e.g; Retain all loci which are covered at-least in two sample by 3 or more reads
meth_cov_filtered = methrix::coverage_filter(m = meth, cov_thr = 3, min_samples = 2, n_threads = 6)
meth_cov_filtered
```

## Subset operations
### Subset by chromosome
```{r}
#Retain sites only from chromosme 1 and "X"
meth_1x = methrix::subset_methrix(m = meth, contigs = c(1, "X"))
meth_1x
```

### Subset by genomic regions
Regions can be data.table or [GRanges](https://kasperdanielhansen.github.io/genbioconductor/html/GenomicRanges_GRanges.html#granges) format.
```{r}
#e.g; Retain sites only in TP53 loci 
tp53_loci = data.table::data.table(chr = "17", start = "7571720", end = "7590868")
meth_tp53 = methrix::subset_methrix(m = meth, regions = tp53_loci)
meth_tp53
```

### Subset by sampels
```{r}
meth_d117 = methrix::subset_methrix(m = meth, samples = "tumor00_JMMLC_D117")
meth_d117
```

## Reorder matrices by sd
```{r}
meth_sd = methrix::order_by_sd(m = meth)
head(methrix::get_matrix(m = meth_sd, type = "M"))
```

## Converting methrix to BSseq
If you prefer to work with [bsseq](https://bioconductor.org/packages/release/bioc/vignettes/bsseq/inst/doc/bsseq.html#3_using_objects_of_class_bsseq) object, you can generate `bsseq` object from methrix with the `methrix2bsseq`. 
```{r}
library(bsseq)
```

```{r}
bs_seq = methrix::methrix2bsseq(m = meth)

bs_seq
```

## SessionInfo
```{r}
sessionInfo()
```

